internal/vm/hcs/vpmem.go

```diff
@@ -0,0 +1,82 @@
+package hcs
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/Microsoft/hcsshim/internal/hcs/resourcepaths"
+	hcsschema "github.com/Microsoft/hcsshim/internal/hcs/schema2"
+	"github.com/Microsoft/hcsshim/internal/requesttype"
+	"github.com/Microsoft/hcsshim/internal/vm"
+	"github.com/pkg/errors"
+)
+
+func (uvmb *utilityVMBuilder) AddVPMemController(maximumDevices uint32, maximumSizeBytes uint64) error {
+	uvmb.doc.VirtualMachine.Devices.VirtualPMem = &hcsschema.VirtualPMemController{
+		MaximumCount:     maximumDevices,
+		MaximumSizeBytes: maximumSizeBytes,
+	}
+	uvmb.doc.VirtualMachine.Devices.VirtualPMem.Devices = make(map[string]hcsschema.VirtualPMemDevice)
+	return nil
+}
+
+func (uvmb *utilityVMBuilder) AddVPMemDevice(ctx context.Context, id uint32, path string, readOnly bool, imageFormat vm.VPMemImageFormat) error {
+	if uvmb.doc.VirtualMachine.Devices.VirtualPMem == nil {
+		return errors.New("VPMem controller has not been added")
+	}
+	imageFormatString, err := getVPMemImageFormatString(imageFormat)
+	if err != nil {
+		return err
+	}
+	uvmb.doc.VirtualMachine.Devices.VirtualPMem.Devices[strconv.Itoa(int(id))] = hcsschema.VirtualPMemDevice{
+		HostPath:    path,
+		ReadOnly:    readOnly,
+		ImageFormat: imageFormatString,
+	}
+	return nil
+}
+
+func (uvmb *utilityVMBuilder) RemoveVPMemDevice(ctx context.Context, id uint32, path string) error {
```

Comment:
@jstarks I'm not a big fan of these, but I didn't want to just rehash the methods that are already on all of the manager interfaces just to leave off remove for creating the document. We don't really have a use case for removing something from the document before sending it to hcs/rpc'ing.

Reply:
Then you should at least return errors.New("not supported") or something.

Reply:
Sure, that works

Reply:
Wanted to see if you had thoughts on reusing the same interfaces on the builder also mainly